How to set private NPM registry for JupyterLab installs?

Hi, a user here. Looked into #7109 but I still have no idea how to implement it ... 
Can someone explain the steps that a user may take to set a private NPM registry for JupyterLab installs?

@fcollonval's any thoughts? It seems as though the default NPM registry gets put into the JupyterLab build. So I'm assuming that setting the registry must take place before building JupyterLab ... but I could be wrong. 

I could not find simple instructions on exactly how to do this anywhere, so that's why I'm filing an issue.

Thanks for the answer. However, I believe this does not work, at least for JupyterLab 1.2.4.
 
Steps I took:
`jlpm config set registry http://my.private.npm.repository --global`
`jupyter labextension install @jupyter-widgets/jupyterlab-manager`
 
Didn't work, out. I also tried this:
`jlpm config set registry http://my.private.npm.repository --global`
`jupyter lab build`
`jupyter labextension install @jupyter-widgets/jupyterlab-manager`
 
I'm on a private network, so I hope that we can find an answer after I type all of this out.  The log files after both of these installs are exactly the same.
 
Brief synopsis, JupyterLab is still using https://registry.npmjs.org as the registry.
 
Log file:
 
```
Yarn Configuration Loaded
Node v10.13.0
 
> H:\Anaconda3\npm.CMD pack @jupyter-widgets/jupyterlab-manager
npm ERR! code ENOTFOUND
npm ERR! erno ENOTFOUND
npm ERR! network request to https://registry.npmjs.org/@jupyter-widgets%2fjupyterlab-manager failed, reason: getaddrinfo ENOTFOUND registry.npmjs.org registry.npmjs.org:443
 
... {condensing ... stuff about bad network connectively & setting the proxy config ... irrelevant}
 
Traceback (most recent call last):
 
    File "H:\Anaconda3\lib\site-packages\jupyterlab\debuglog.py", line 47, in debug_logging
        yield
 
    File "H:\Anaconda3\lib\site-packages\jupyterlab\labextensions.py", line 93, in start
        ans = self.run_task()
 
    File "H:\Anaconda3\lib\site-packages\jupyterlab\labextensions.py", line 149, in run_task
        for i, arg in enumerate(self.extra_args)
 
    File "H:\Anaconda3\lib\site-packages\jupyterlab\labextensions.py", line 149, in <listcomp>
        for i, arg in enumerate(self.extra_args)
 
    File "H:\Anaconda3\lib\site-packages\jupyterlab\commands.py", line 398, in install_extension
        return handler.Install_extension(extension, pin=pin)
 
    File "H:\Anaconda3\lib\site-packages\jupyterlab\commands.py", line 611, in install_extension
        info = self._install_extension(extension, tempdir, pin=pin)
 
    File "H:\Anaconda3\lib\site-packages\jupyterlab\commands.py", line 1501, in _install_extension
        info = self._extract_package(extension, tempdir, pin=pin)
 
    File "H:\Anaconda3\lib\site-packages\jupyterlab\commands.py", line 1571, in _extract_package
        raise ValueError (msg % source)
```

Also, the output to `jlpm config list` shows that my registry is set to http://my.private.npm.repository/ (both from the NPM config & the yarn config)

--- a/docs/source/getting_started/installation.rst
+++ b/docs/source/getting_started/installation.rst
@@ -121,7 +121,7 @@ Usage with private NPM registry
 
 To install extensions, you will need access to a NPM packages registry. Some companies do not allow
 reaching directly public registry and have a private registry. To use it, you need to configure ``npm``
-and ``yarn`` to point to that registry (ask you corporate IT department for the correct URL):
+**and** ``yarn`` to point to that registry (ask your corporate IT department for the correct URL):
 
 .. code::
 

--- a/docs/source/getting_started/installation.rst
+++ b/docs/source/getting_started/installation.rst
@@ -116,6 +116,25 @@ CSS Variables, and are not directly supported at this time.
 A tool like `postcss <https://postcss.org/>`__ can be used to convert the CSS files in the
 ``jupyterlab/build`` directory manually if desired.
 
+Usage with private NPM registry
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To install extensions, you will need access to a NPM packages registry. Some companies do not allow
+reaching directly public registry and have a private registry. To use it, you need to configure ``npm``
+and ``yarn`` to point to that registry (ask you corporate IT department for the correct URL):
+
+.. code::
+
+    npm config set registry https://registry.company.com/
+    yarn config set registry https://registry.company.com/
+    
+JupyterLab will pick up that registry automatically.
+
+.. note::
+
+    You can check which registry URL is used by JupyterLab by running::
+    
+      python -c "from jupyterlab.commands import AppOptions; print(AppOptions().registry)"
 
 Installation problems
 ~~~~~~~~~~~~~~~~~~~~~
@@ -182,7 +201,7 @@ which means, SSL communication will not be required for downloading Python libra
 Using the tips from above, you can handle many network problems 
 related to installing Python libraries.
 
-Many Jupyter extensions require having a working ``npm`` command,
+Many Jupyter extensions require having a working ``npm`` and ``jlpm`` (alias for ``yarn``) commands,
 which is required for downloading useful Jupyter extensions or other JavaScript dependencies.
 
 Example of typical error message, when ``npm`` cannot connect to own repositories:
@@ -197,6 +216,7 @@ Example of typical error message, when ``npm`` cannot connect to own repositorie
 
     # Set default registry for NPM (optional, useful in case if common JavaScript libs cannot be found)
     npm config set registry http://registry.npmjs.org/
+    jlpm config set registry https://registry.yarnpkg.com/
 
 
 In case you can communicate via HTTP, but installation with ``npm`` fails

